src/lib.rs

```diff
@@ -1277,6 +1277,44 @@ impl Error {
         self.depth
     }
 
+    /// Inspect the underlying [`io::Error`] if there is one.
+    /// If a cycle was detected instead, [`None`] is returned.
+    ///
+    /// ```rust,no-run
+    /// use walkdir::{WalkDir};
+    /// use std::io;
+    /// use std::path::Path;
+    ///
+    /// let mut it = WalkDir::new("foo").into_iter();
+    /// loop {
+    ///     let entry = match it.next() {
+    ///         None => break,
+    ///         Some(Ok(entry)) => entry,
+    ///         Some(Err(err)) => {
+    ///             println!("failed to access entry {}", err.path().unwrap_or(Path::new("")).display());
+    ///             if let Some(inner) = err.io_error() {
+    ///                 match inner.kind() {
+    ///                     io::ErrorKind::InvalidData => println!("entry contains invalid data: {}", inner),
+    ///                     io::ErrorKind::PermissionDenied => println!("Missing permission to read entry: {}", inner),
+    ///                     _ => println!("Unexpected error occurred: {}", inner),
+    ///                 }
+    ///             }
+    ///             continue
+    ///         }
+    ///     };
+    ///     println!("{}", entry.path().display());
+    /// }
+    /// ```
+    ///
+    /// [`None`]: https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.None
+    /// [`io::Error`]: https://doc.rust-lang.org/stable/std/io/struct.Error.html
+    pub fn io_error(&self) -> Option<&io::Error> {
```

Comment:
I wonder if this could be a `Cow<'a, io::Error>`? Then you could always return an error, but if it's not already an `io::Error` then you'll get `Cow::Owned`.

You should just be able to match on `err.io_error().kind()` then.

Reply:
That would be great, except for the fact that to do that, `io::Error` needs to implement `ToOwned`, which it doesn't.

Reply:
Ah right. That's a bit of a shame